src/core/featuremodel.cpp

```diff
@@ -46,15 +46,26 @@ void FeatureModel::setCurrentLayer( QgsVectorLayer* layer )
   if ( layer == mLayer )
     return;
 
+  //remember the last feature of the old layer
```

Comment:
Do we need something like this?

```
slot FeatureModel::onLayerRemoved( layer )
{
  mRememberings.remove( layer );
}

connect ( mLayer, destroyed, ..., Qt::UniqueConnection );
```

Reply:
I made it now in a different way. After project is loaded I reset the map. I liked your approach with the layers, but when the model never is destroyed the layer is never destroyed as well, right?

Reply:
The layer should be destroyed on project unload, it's owned by the project, not by this model.
To be honest, I am not a fan of giving the responsibility for clearing this to QML code.

Reply:
you are right. than it's even simpler. thanks.

Reply:
@m-kuhn Probably you have a quick input for my problem because I guess you used destroyed signals before, otherwise tell me to investigate more myself, but I'm feeling to loose time here:
```
void FeatureModel::removeLayer( QObject* layer )
{
  mRememberings.remove( qobject_cast< QgsVectorLayer * >( layer ) );
}

connect( mLayer, &QgsVectorLayer::destroyed, this, &FeatureModel::removeLayer );
```
This does not remove the layer from the map. I don't know why. It's hard to find out in the debugger what `layer` does contain. `mRememberings.contains( qobject_cast< QgsVectorLayer * >( layer ) ) ` is false as well. Means the  `layer`  is not the  `mLayer` we stored in remember map.

It's possible that this is because the derived object (QgsVectorLayer) is actually destroyed before the signal is emitted. But not sure. I tried another possiblity, hoping it keeps the layer, but it's not even arriving in the slot (possibly not emitting the signal:
```
void FeatureModel::removeLayer( QgsVectorLayer* layer )
{
  mRememberings.remove( layer );
}
  connect( mLayer, SIGNAL(destroyed(QgsVectorLayer*)), this, SLOT(removeLayer(QgsVectorLayer*)));
```

Reply:
Seems to work with `static_cast`, thanks.
If I do understand it correctly, the `qobject_cast` only gives an object, if there as been a proper cast and `static_cast` gives us just a converted zombie-object that is usually not usable at all but for this case it is, because the pointer address stays the same. Correct?

Reply:
`qobject_cast` is pretty much like `dynamic_cast`, they do actually check (at runtime) if the object can be converted to a given subclass.

`static_cast` checks at build time if this is something you can do, but doesn't do any checks at runtime.

E.g.

```
QObject *myobject = new QWidget();

// this is returning a nullptr, because it looks at myobject and realizes that myobject is actually a QWidget which cannot be converted to a QgsVectorLayer
QgsVectorLayer *layer = qobject_cast<QgsVectorLayer*>(myobject);

// QgsVectorLayer is a subclass of QObject, so static_cast will just do the conversion. Even if in this case it's a Widget which is not safe to cast to a QgsVectorLayer. Any attempt to use layer-specific methods on QgsVectorLayer result in undefined behavior (in most cases a crash)
QgsVectorLayer *layer = static_cast<QgsVectorLayer*>(myobject);
```

Reply:
Thanks @m-kuhn